feat: add openharmony header

richerfu opened the pull request:
This is a new PR #20020, which adds OpenHarmony header files for use with zig cc. Following the suggestions from PR #20020 and #21809, this PR only includes the logic for generating header files, while the remaining adaptation logic will be handled in other PRs.

cc @alexrp 

Diff logic:
1. generate the full musl headers map and key is relative path:
  i. Put musl/include
  ii. Put musl/arch/generic
  iii. Put musl/arch/$arch
2. iterate the full musl headers map and check if the file exists in the search path:
  i. If the file exists, read the file and remove the comments and spaces and lines
  ii. Calculate the hash of the file content and check if the hash is the same
  iii. If the hash is the same, we can reduce the size
  iv. If the hash is different, we should save the content
3. Save the content to the out dir
  i. If the hit count is greater than 1, save the content to the generic-ohos dir
  ii. If the hit count is equal to 1, save the content to the target dir

richerfu committed:
feat: add script

richerfu committed:
feat: add ohos headers

alexrp commented:
So the idea is to use the original musl headers for cases where they are semantically identical to the OpenHarmony ones anyway? Good idea, I like it.

A few thoughts:

* `arm-linux-ohos` should be `arm-linux-ohoseabi` (see #21605).
* It seems like all the header files added in this PR lost their trailing newline.
* I don't understand why e.g. `aarch64-linux-ohos/bits/stat.h` shows up in this PR, since it seems to be identical to `generic-musl/bits/stat.h`.
* I wonder if the `process_headers_ohos.zig` script could be merged with `process_headers.zig` now, with an `--abi ohos` flag? The latter script has been simplified quite a bit since your first PR (and will be simplified more in #22090), so this might make sense to do.

richerfu committed:
feat: add new line & rename arm-linux-ohos to arm-linux-ohoseabi

richerfu commented:
According to https://github.com/ziglang/zig/blob/master/src/musl.zig#L461 , it seems that when the musl is built, the original musl header file will be searched first.
Therefore, the new header file only provides the parts that are different from musl.

- `arm-linux-ohoseabi` There are no differences from the original musl's arm architecture, so no additional files will be generated
- `aarch64-linux-ohos/bits/stat.h` There is a very small difference.
   ```c
   struct stat {
	dev_t st_dev;
	ino_t st_ino;
	mode_t st_mode;
	nlink_t st_nlink;
	uid_t st_uid;
	gid_t st_gid;
	dev_t st_rdev;
	unsigned long __pad;
	off_t st_size;
	blksize_t st_blksize;
	int __pad2;
	blkcnt_t st_blocks;
	struct timespec st_atim;
	struct timespec st_mtim;
	struct timespec st_ctim;
        // the musl's origin code is unsigned __unused[2] and ohos's file one char is more.
	unsigned __unused1[2];
   }; 

- I think merging is not very good, because in the process of comparing the differences between ohos header files, a lot of processing is not needed by the original script, and we rely on the musl header files that have been processed by the original script. Independent scripts help us maintain musl and ohos versions of musl more conveniently, and their code will look much simpler.

alexrp commented:
> According to https://github.com/ziglang/zig/blob/master/src/musl.zig#L461 , it seems that when the musl is built, the original musl header file will be searched first.
> Therefore, the new header file only provides the parts that are different from musl.

Note that this code is for building static musl artifacts. Since, as we discussed previously, we will only support dynamic linking for OpenHarmony, this code only matters for building `crti.o`/`crtn.o` and `crt1.o`/`scrt1.o`.

For the include paths that are exposed to user code, see:

https://github.com/ziglang/zig/blob/aa7d138462602e086aacf738e4b92bfa3372bebe/lib/std/zig/LibCDirs.zig#L169-L214

In both code paths, we will need to check for `target.abi.isOpenHarmony()` and add the OpenHarmony paths before the regular musl paths so that the headers that are different will be picked up as expected.

> `aarch64-linux-ohos/bits/stat.h` There is a very small difference.

Ah, ok. That's pretty silly, but not easy to fix without hacky special cases in the script.

alexrp commented:
> I think merging is not very good, because in the process of comparing the differences between ohos header files, a lot of processing is not needed by the original script, and we rely on the musl header files that have been processed by the original script. Independent scripts help us maintain musl and ohos versions of musl more conveniently, and their code will look much simpler.

In any case, I suggest getting rid of all that `MultiArch`/`MultiAbi` complexity. I removed that from `process_headers.zig` here: https://github.com/ziglang/zig/pull/21717/files#diff-ac8b4b89c2c051974a6382b5268835c94ce84ce19e955ff2568c79fbcfd8e0c0

richerfu committed:
Remove MultiArch and MultiAbi

richerfu commented:
> > I think merging is not very good, because in the process of comparing the differences between ohos header files, a lot of processing is not needed by the original script, and we rely on the musl header files that have been processed by the original script. Independent scripts help us maintain musl and ohos versions of musl more conveniently, and their code will look much simpler.
> 
> In any case, I suggest getting rid of all that `MultiArch`/`MultiAbi` complexity. I removed that from `process_headers.zig` here: https://github.com/ziglang/zig/pull/21717/files#diff-ac8b4b89c2c051974a6382b5268835c94ce84ce19e955ff2568c79fbcfd8e0c0

Removed.

richerfu commented:
> That's pretty silly, but not easy to fix without hacky special cases in the script.

I'm not sure if changing the field name will have any impact on the system or code. So I'm going to keep the difference.

andrewrk commented on .gitignore:
```diff
@@ -16,6 +16,7 @@ zig-out/
 /build/
 /build-*/
 /docgen_tmp/
+.DS_Store
```
If you can't be bothered to read the comment above, why should I trust your other work in this PR?

richerfu replied:
I'm sorry I overlooked the description about `gitignore`, I have rolled back the code.

richerfu committed:
revert gitignore config

richerfu committed:
Merge branch 'master' into feat-openharmony-header

richerfu committed:
feat: add basic adaption for ohos

alexrp commented:
Discussed continued at https://codeberg.org/ziglang/zig/pulls/31637.
